### DISCUSSION TITLE
Why timeout is an int pointer?

### DISCUSSION THREAD

**Discussion by fedor-beznravny** [moby/moby#45033] [Q&A] [answered]
Hello!

Can someone suggest, why [here](https://github.com/moby/moby/blob/v23.0.1/api/types/container/config.go#L32) timeout is requiered as pointer to int?

The same sitation is visible in other places too.

Pointer can accept nil as "use default value" notifier. But, what if I want to specify 3 seconds - I should write `&int(3)`?

Another question - why StopOptions are mandatory, why, as in official example, we could not pass a nil?

**Comment by fedor-beznravny** [comment]
Problem is in data lifecycle? Manual allocation of heap data and simple stack value copy is equal from the result point of view.

**Comment by thaJeztah** [comment, marked as answer]
The pointer is used to distinguish between "no value set" (`nil`) and `0`, which is a bit of a workaround around how Golang handles (un)marshalling JSON. For non-pointer values, Golang doesn't distinguish between a field not being set and fields that are set, but set to their default (`0` for integers). In this case, `0` was accepted as a valid value (don't wait), so `nil` is used as way to detect that no custom value was set (so the default value should be used);

https://github.com/moby/moby/blob/v23.0.1/api/types/container/config.go#L24-L32
